### PR TITLE
Migrate fixup code to use agent logic

### PR DESCRIPTION
**Pull request by pkukielka**
Fixes CODY-2802

Corresponding Cody PR: https://github.com/sourcegraph/cody/pull/5019

## Changes

This PR removed a lot of custom logic around fixups and instead relies on a agent implementation.
All interactions with client are done either by lenses API (like informing JetBrains which actions it should show to user) or through edits API (to notify JetBrains which text modifications it should apply). JetBrains in turn is able to signal few fixup specific events back to agent (like `accept`, `retry`, or `cancel`). That's it.

It's worth noting that right now all of the `com.sourcegraph.cody.edit.actions.lenses` actions are specific to fixups, but nothing is blocking us from using the same mechanism for other features which utilises lenses.

Specific changes:

1. `FixupSession` and `FixupService` are completely removed. JetBrains implementation does not duplicate fixups logic anymore.
2. `LensesService` was added - it is responsible for receiving and rendering lenses provided by agent.
Those lenses often have commands assigned, see [cody: items.ts](https://github.com/sourcegraph/cody/blob/main/vscode/src/non-stop/codelenses/items.ts)
JetBrains implementation is expected to have actions with the matching IDs - in such case they will be wired automatically and click when user interact with the lenses. That is also a reason of some renames in `plugin.xml`.
3. `EditService` was added - it is responsible for receiving and applying [TextEdit](https://code.visualstudio.com/api/references/vscode-api#TextEdit) and [WorkspaceEdit](https://code.visualstudio.com/api/references/vscode-api#WorkspaceEdit) events from agent. Those edits are representing low level editor edits in VSC and have no direct relation with fixups or any other high level constructs. That is a major change compared to our previous implementation.

## Test plan

1. Full QA will be needed on those changes. It need to be tested with the CODY_DIR set to [this PR](https://github.com/sourcegraph/cody/pull/5019)

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/edit/actions/lenses/LensEditAction.kt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  override fun actionPerformed(e: AnActionEvent) {
+    try {
+      var project = e.project
+      if (project == null) {
+        project = e.dataContext.getData(PlatformDataKeys.PROJECT.name) as? Project
+      }
+      if (project == null || project.isDisposed) {
+        logger.warn("Received code lens action for null or disposed project: $e")
```
Comment (odisseus) [inline src/main/kotlin/com/sourcegraph/cody/edit/actions/lenses/LensEditAction.kt:41]: This probably should be an error.

File: src/main/resources/CodyBundle.properties [hunk 2]
```diff
@@ -185,6 +185,7 @@ filter.sidebar-panel-ignored-file.text=<html>This file has been restricted by an
 filter.sidebar-panel-ignored-file.learn-more-cta=Learn more
 
 # Other Actions
+action.cody.not-working=Cody is disabled or still starting up
```
Comment (odisseus) [inline src/main/resources/CodyBundle.properties:188]: We already have messages for `notification.general.cody-not-running.title` and `notification.general.cody-not-running.detail`. Does this new message have a different use case?

**Review by odisseus — APPROVED** [review]
Generally looks good

File: src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    listeners.remove(listener)
+  }
+
+  fun updateLenses(uri: String, codeLens: List<ProtocolCodeLens>) {
+    runInEdt {
+      val vf = CodyEditorUtil.findFileOrScratch(project, uri) ?: return@runInEdt
+      val fileDesc = OpenFileDescriptor(project, vf)
+      val editor =
+          FileEditorManager.getInstance(project).openTextEditor(fileDesc, true) ?: return@runInEdt
+
```
Comment (RXminuS) [inline src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt:66]: Maybe some assertion that there's never partially overlapping lens ranges. Just in case

File: src/main/resources/META-INF/plugin.xml [hunk 4]
```diff
@@ -240,72 +240,59 @@
...
-                <action id="cody.editShowDiffAction"
-                        class="com.sourcegraph.cody.edit.EditShowDiffAction"
+                <action id="cody.fixup.codelens.diff"
+                        class="com.sourcegraph.cody.edit.actions.lenses.EditShowDiffAction"
                         text="Show Diff"
                         description="Show a diff view of the fixup">
                     <keyboard-shortcut replace-all="true" first-keystroke="ctrl alt D" keymap="$default"/>
                     <keyboard-shortcut replace-all="true" first-keystroke="ctrl alt K" keymap="Mac OS X 10.5+"/>
                     <override-text place="GoToAction" text="Cody: Show Diff"/>
                 </action>
```
Comment (RXminuS) [inline src/main/resources/META-INF/plugin.xml]: There seems to be some conflicts with existing shortcuts. We'll check with @danielmarquespt today :-)

File: src/main/kotlin/com/sourcegraph/cody/chat/ui/LlmDropdown.kt [hunk 5]
```diff
@@ -73,7 +74,7 @@ class LlmDropdown(
     val selectedFromState = chatModelProviderFromState
     val selectedFromHistory = HistoryService.getInstance(project).getDefaultLlm()
     val selectedModel =
```
Comment (RXminuS) [inline src/main/kotlin/com/sourcegraph/cody/chat/ui/LlmDropdown.kt:76]: Double check if we need to handle nulls (during init)

File: src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentServer.kt [hunk 6]
```diff
@@ -125,8 +124,11 @@ interface _LegacyAgentServer {
 
   @JsonRequest("editTask/cancel") fun cancelEditTask(params: TaskIdParam): CompletableFuture<Void?>
 
-  @JsonRequest("editTask/getFoldingRanges")
-  fun getFoldingRanges(params: GetFoldingRangeParams): CompletableFuture<GetFoldingRangeResult>
+  @JsonRequest("editTask/retry")
+  fun retryEditTask(params: EditTask_RetryParams): CompletableFuture<EditTask>
+
+  @JsonRequest("editTask/getTaskDetails")
+  fun getEditTaskDetails(params: TaskIdParam): CompletableFuture<EditTask>
```
Comment (RXminuS) [inline src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentServer.kt]: Move the to the AutoGenerated subset. That way it'll be easy to keep track of which methods have been "migrated" and which ones are pending

**Review by RXminuS — DISMISSED** [review]

File: src/integrationTest/kotlin/com/sourcegraph/cody/edit/DocumentCodeTest.kt [hunk 7]
```diff
@@ -1,163 +1,114 @@
...
-    // A more robust check is to see if the selection "range" is just the caret position.
-    // If so, then our fallback range somehow made the round trip, which is bad. The lenses will
-    // go in the wrong places, etc.
-    val caret = runInEdtAndGet { myFixture.editor.caretModel.primaryCaret.offset }
-    assertFalse(
-        "Selection range should not equal the caret position",
-        selection.startOffset == caret && selection.endOffset == caret)
-  }
-
-  @Ignore
```
Comment (mkondratek) [inline src/integrationTest/kotlin/com/sourcegraph/cody/edit/DocumentCodeTest.kt:48]: wow - can we reenable this test now? awesome 🚀

File: src/integrationTest/kotlin/com/sourcegraph/cody/edit/DocumentCodeTest.kt [hunk 8]
```diff
@@ -1,163 +1,114 @@
...
+import org.hamcrest.Matchers.startsWith
+import org.junit.Assert.assertThat
 import org.junit.Test
 import org.junit.runner.RunWith
 
 @RunWith(CustomJunitClassRunner::class)
 class DocumentCodeTest : CodyIntegrationTextFixture() {
-
-  @Test
-  fun testGetsFoldingRanges() {
```
Comment (mkondratek) [inline src/integrationTest/kotlin/com/sourcegraph/cody/edit/DocumentCodeTest.kt:30]: folding ranges are handled by cody already, is that right?

File: src/main/java/com/sourcegraph/cody/agent/CodyAgentClient.java [hunk 9]
```diff
@@ -3,6 +3,8 @@
 import com.intellij.openapi.application.ApplicationManager;
 import com.intellij.openapi.diagnostic.Logger;
 import com.sourcegraph.cody.agent.protocol.*;
+import com.sourcegraph.cody.agent.protocol_generated.DisplayCodeLensParams;
```
Comment (mkondratek) [inline src/main/java/com/sourcegraph/cody/agent/CodyAgentClient.java:6]: I see no `DisplayCodeLensParams.kt` 🫤 I guess the protocol should be re-generated. is that right?

File: src/main/kotlin/com/sourcegraph/cody/edit/FixupService.kt [hunk 10]
```diff
@@ -1,177 +0,0 @@
...
-
-  fun notifySessionStateChanged() {
-    synchronized(fixupSessionStateListeners) {
-      for (listener in fixupSessionStateListeners) {
-        listener.fixupSessionStateChanged(isEditInProgress())
-      }
-    }
-  }
-
-  interface ActiveFixupSessionStateListener {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/edit/FixupService.kt:167]: so nice we can drop this one in particular 😌

File: src/test/kotlin/com/sourcegraph/cody/KeyboardShortcutTest.kt [hunk 11]
```diff
@@ -28,11 +28,8 @@ class KeyboardShortcutTest : BasePlatformTestCase() {
             "cody.editCodeAction" to arrayOf("ctrl alt COMMA", "ctrl alt ENTER"),
             "cody.documentCodeAction" to arrayOf("ctrl alt H", "ctrl alt H"),
             "cody.testCodeAction" to arrayOf("ctrl alt G", "ctrl alt G"),
-
-            // Handles cody.inlineEditUndoAction, inlineEditCancelAction, & inlineEditDismissAction:
-            "cody.editCancelOrUndoAction" to arrayOf("ctrl alt BACK_SPACE", "ctrl alt BACK_SPACE"),
             "cody.editShowDiffAction" to arrayOf("ctrl alt D", "ctrl alt K"),
-            "cody.inlineEditAcceptAction" to arrayOf("ctrl shift EQUALS", "ctrl shift EQUALS"),
+            "cody.fixup.codelens.accept" to arrayOf("ctrl shift EQUALS", "ctrl shift EQUALS"),
```
Comment (mkondratek) [inline src/test/kotlin/com/sourcegraph/cody/KeyboardShortcutTest.kt]: the action ids seem to be inconsistent now 😕

File: src/test/kotlin/com/sourcegraph/cody/KeyboardShortcutTest.kt [hunk 12]
```diff
@@ -28,11 +28,8 @@ class KeyboardShortcutTest : BasePlatformTestCase() {
             "cody.editCodeAction" to arrayOf("ctrl alt COMMA", "ctrl alt ENTER"),
             "cody.documentCodeAction" to arrayOf("ctrl alt H", "ctrl alt H"),
             "cody.testCodeAction" to arrayOf("ctrl alt G", "ctrl alt G"),
-
-            // Handles cody.inlineEditUndoAction, inlineEditCancelAction, & inlineEditDismissAction:
-            "cody.editCancelOrUndoAction" to arrayOf("ctrl alt BACK_SPACE", "ctrl alt BACK_SPACE"),
             "cody.editShowDiffAction" to arrayOf("ctrl alt D", "ctrl alt K"),
```
Comment (mkondratek) [inline src/test/kotlin/com/sourcegraph/cody/KeyboardShortcutTest.kt]: whoops - looks like a mistake - shouldnt it be `"cody.fixup.codelens.diff"` now?

File: src/main/kotlin/com/sourcegraph/cody/edit/EditCommandPrompt.kt [hunk 13]
```diff
@@ -611,30 +603,15 @@ class EditCommandPrompt(
     /** Returns a compact symbol representation of the action's keyboard shortcut, if any. */
     @JvmStatic
     fun getShortcutDisplayString(actionId: String): String? {
-      fun getFirstShortcut(id: String): String? {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/edit/EditCommandPrompt.kt:614]: very nice 🥇

File: src/main/kotlin/com/sourcegraph/cody/config/migration/ChatTagsLlmMigration.kt [hunk 14]
```diff
@@ -40,7 +40,7 @@ object ChatTagsLlmMigration {
   }
 
   fun addTagIf(llm: LLMState, tag: String, condition: Boolean?) {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/config/migration/ChatTagsLlmMigration.kt]: can be private

File: src/main/kotlin/com/sourcegraph/cody/edit/actions/lenses/EditShowDiffAction.kt [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+                editTask.selectionRange.start.toOffset(documentBefore),
+                editTask.selectionRange.end.toOffset(documentBefore),
+                editTask.originalText ?: "")
+            ShowDocumentDiffAction(documentBefore, documentAfter).actionPerformed(event)
+          }
+        }
+      }
+    }) {
+  companion object {
+    const val ID = "cody.fixup.codelens.diff"
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/edit/actions/lenses/EditShowDiffAction.kt:29]: we have two ids that can go different easily (here and in plugin xml). but it can be the best we can do

File: src/main/kotlin/com/sourcegraph/cody/edit/widget/LensWidgetGroup.kt [hunk 16]
```diff
@@ -89,12 +89,7 @@ class LensWidgetGroup(val session: FixupSession, parentComponent: Editor) :
 
   private var lastComputedIndent = RECOMPUTE
 
-  var isInWorkingGroup = false
-  var isAcceptGroup = false
-  var isErrorGroup = false
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/edit/widget/LensWidgetGroup.kt:94]: yes, please 🙏

File: src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+      val newLensGroups =
+          codeLens
+              .groupBy { it.range }
+              .map { (range, codeLensesModels) ->
+                val taskId = codeLensesModels.firstNotNullOf { getTaskId(it) }
+                taskId to (range to createLensGroup(editor, codeLensesModels))
+              }
+              .toMap()
+
+      lensGroups[uri]?.values?.forEach { it.second.dispose() }
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt]: The Jetbrains documentation [says](https://plugins.jetbrains.com/docs/intellij/disposers.html#diagnosing-disposer-leaks) that Disposable.dispose() should never be called directly

File: src/main/resources/META-INF/plugin.xml [hunk 18]
```diff
@@ -231,7 +231,7 @@
 
                 <action id="cody.editCodeAction"
```
Comment (mkondratek) [inline src/main/resources/META-INF/plugin.xml:232]: rename here too?

File: src/main/kotlin/com/sourcegraph/common/ShowDocumentDiffAction.kt [hunk 19]
```diff
@@ -44,13 +41,9 @@ class EditShowDiffAction : CompareFileWithEditorAction() {
           editorFile == null -> "Cody Diff"
           else -> "Cody Diff: $editorContentTitle"
         }
-    chain.title1 = "Before Cody Inline Edit"
+    chain.title1 = "Before"
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/common/ShowDocumentDiffAction.kt:44]: why this change?

File: src/main/kotlin/com/sourcegraph/cody/chat/ui/LlmDropdown.kt [hunk 20]
```diff
@@ -27,6 +27,7 @@ class LlmDropdown(
     private val onSetSelectedItem: (ChatModelsResponse.ChatModelProvider) -> Unit,
     val parentDialog: EditCommandPrompt?,
     val chatModelProviderFromState: ChatModelsResponse.ChatModelProvider?,
+    val model: String? = null
 ) : ComboBox<ChatModelsResponse.ChatModelProvider>(MutableCollectionComboBoxModel()) {
   var hasServerSentModels = false
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/chat/ui/LlmDropdown.kt]: nit: this should be private probably

File: src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+import com.sourcegraph.cody.edit.widget.LensLabel
+import com.sourcegraph.cody.edit.widget.LensSpinner
+import com.sourcegraph.cody.edit.widget.LensWidgetGroup
+import com.sourcegraph.utils.CodyEditorUtil
+import javax.swing.Icon
+
+interface LensListener {
+  fun onLensesUpdate(
+      uri: String,
+      taskId: String?,
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt]: unused I guess

File: src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+import com.sourcegraph.cody.edit.widget.LensIcon
+import com.sourcegraph.cody.edit.widget.LensLabel
+import com.sourcegraph.cody.edit.widget.LensSpinner
+import com.sourcegraph.cody.edit.widget.LensWidgetGroup
+import com.sourcegraph.utils.CodyEditorUtil
+import javax.swing.Icon
+
+interface LensListener {
+  fun onLensesUpdate(
+      uri: String,
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt]: unused too

File: src/main/kotlin/com/sourcegraph/cody/edit/EditCommandPrompt.kt [hunk 23]
```diff
@@ -74,18 +76,20 @@ import javax.swing.event.CaretListener
 
 /** Pop up a user interface for giving Cody instructions to fix up code at the cursor. */
 class EditCommandPrompt(
-    val controller: FixupService,
+    val project: Project,
     val editor: Editor,
     dialogTitle: String,
-    instruction: String? = null
-) : JFrame(), Disposable, FixupService.ActiveFixupSessionStateListener, DataProvider {
+    val previousEdit: EditTask? = null
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/edit/EditCommandPrompt.kt]: private?

File: src/main/kotlin/com/sourcegraph/cody/edit/actions/lenses/LensEditAction.kt (hunk 1 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/edit/actions/lenses/LensEditAction.kt:41]: Not really, it can happen if user do an edit and close project at the same time

File: src/main/kotlin/com/sourcegraph/cody/chat/ui/LlmDropdown.kt (hunk 5 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/chat/ui/LlmDropdown.kt:76]: I wanted to add additional check but `it.model` cannot be null so IDE was saying that check was redundant.

File: src/integrationTest/kotlin/com/sourcegraph/cody/edit/DocumentCodeTest.kt (hunk 8 above)
Comment (pkukielka) [reply src/integrationTest/kotlin/com/sourcegraph/cody/edit/DocumentCodeTest.kt:30]: yep, we do not need them at all

File: src/main/java/com/sourcegraph/cody/agent/CodyAgentClient.java (hunk 9 above)
Comment (pkukielka) [reply src/main/java/com/sourcegraph/cody/agent/CodyAgentClient.java:6]: It's checked-in already thanks to previous PR from @RXminuS 
https://github.com/sourcegraph/cody/blob/main/agent/bindings/kotlin/lib/src/main/kotlin/com/sourcegraph/cody/agent/protocol_generated/DisplayCodeLensParams.kt

File: src/test/kotlin/com/sourcegraph/cody/KeyboardShortcutTest.kt (hunk 11 above)
Comment (pkukielka) [reply src/test/kotlin/com/sourcegraph/cody/KeyboardShortcutTest.kt]: I cannot help it, unless I want to change all of them.
I think we should try to make them the same as in cody where it makes sense :)
Here I need that for auto-binding without custom mapping one ID to the other (which I think is a way to go).

File: src/main/kotlin/com/sourcegraph/common/ShowDocumentDiffAction.kt (hunk 19 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/common/ShowDocumentDiffAction.kt:44]: Because I moved it to common ;) It does not have to be specific to inline edit although. it is now.
TBH I think it that change does not hurt it too much.

File: src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt (hunk 22 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt]: yea, I will clean it up

File: src/test/kotlin/com/sourcegraph/cody/KeyboardShortcutTest.kt (hunk 12 above)
Comment (pkukielka) [reply src/test/kotlin/com/sourcegraph/cody/KeyboardShortcutTest.kt]: Good catch!

File: src/main/resources/META-INF/plugin.xml (hunk 18 above)
Comment (pkukielka) [reply src/main/resources/META-INF/plugin.xml:232]: Not needed, as this is not called from agent, and there is one good reason to not do renames if not needed:
I think it can mess-up shortcuts if users already assigned custom ones to their actions.

**Review by RXminuS — COMMENTED** [review]
Where are we on this?

File: src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt (hunk 3 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt:66]: Added!

File: src/main/resources/META-INF/plugin.xml (hunk 4 above)
Comment (pkukielka) [reply src/main/resources/META-INF/plugin.xml]: Fixed. I consulted changes with @danielmarquespt

File: src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentServer.kt (hunk 6 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentServer.kt]: Done!

File: src/main/kotlin/com/sourcegraph/cody/chat/ui/LlmDropdown.kt (hunk 20 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/chat/ui/LlmDropdown.kt]: Done

File: src/main/kotlin/com/sourcegraph/cody/edit/EditCommandPrompt.kt (hunk 23 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/edit/EditCommandPrompt.kt]: Done

**Comment by pkukielka** [comment]
> Where are we on this?

@RXminuS It's ready for re-review! All remarks should be fixed.

**Comment by RXminuS** [comment]
* **#1921** <a href="https://app.graphite.dev/github/pr/sourcegraph/jetbrains/1921?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1955** <a href="https://app.graphite.dev/github/pr/sourcegraph/jetbrains/1955?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @pkukielka and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by mkondratek — APPROVED** [review]
